crates/agent/src/publications/linked_materializations.rs

```diff
@@ -0,0 +1,397 @@
+                None
+            }
+        })
+        .collect::<std::collections::HashSet<_>>();
+    maybe_linked_captures.into_iter().collect()
+}
+
+/// Validates the `sourceCapture` field of any materializations in the draft.
+/// It ensures that the materialization spec is read-authorized to the prefix
+/// of the `sourceCapture`, and that it exists and is actually a capture.
+/// For example, given `sourceCapture: acmeCo/foo/bar`, we'd require that the
+/// materialization spec be read authorized to all of `acmeCo/foo/`. This is
+/// because the capture may discover new collections, which would be put under
+/// that prefix. This is not strictly required for correctness, since we'd still
+/// enforce authZ when the materialization is published with the new bindings.
+/// But without this check, we might leak information about collections that
+/// a user doesn't have access to, since they'd be able to see them in the
+/// draft_specs.
+pub async fn validate_source_captures(
+    txn: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+    draft_catalog: &models::Catalog,
+    spec_rows: &[SpecRow],
+) -> anyhow::Result<Vec<Error>> {
+    let mut errors = Vec::new();
+    let mut other_checks = Vec::new();
+
+    for row in spec_rows
+        .iter()
+        .filter(|r| r.draft_type == Some(CatalogType::Materialization))
+    {
+        let spec = draft_catalog
+            .materializations
+            .get(&models::Materialization::new(row.catalog_name.as_str()))
+            .expect("draft catalog must contain items from spec_rows");
+        let Some(source_capture) = spec.source_capture.as_ref() else {
+            continue;
+        };
+        let starting_error_count = errors.len();
+        let catalog_name = row.catalog_name.as_str();
+        let spec_capabilities = &row.spec_capabilities;
+        // Check the permissions before checking that the sourceCapture exists
+        // and is a capture, so we avoid leaking information about unauthorized
+        // entities.
+        if let Some(last_slash_index) = source_capture.rfind('/') {
+            let capture_prefix = &source_capture[..(last_slash_index + 1)];
+            if !spec_capabilities.iter().any(|cap| {
+                capture_prefix.starts_with(&cap.object_role) && cap.capability >= Capability::Read
+            }) {
+                errors.push(Error {
+                    catalog_name: row.catalog_name.clone(),
+                    detail: format!(
+                        "Specification '{catalog_name}' is not read-authorized to `sourceCapture` prefix: '{capture_prefix}'.\nAvailable grants are: {}",
+                        serde_json::to_string_pretty(&spec_capabilities.0).unwrap(),
+                    ),
+                    ..Default::default()
+                });
+            }
+        } else {
+            // The value of sourceCapture is invalid, since all catalog names must include
+            // at least one "/"
+            errors.push(Error {
```

Comment:
The DB constraints don't help us here because this value is taken from the `spec`, which doesn't have a separate constraint on `spec->>'sourceCapture'`. We could technically remove this and just consider it a "sourceCapture does not exist" error, but I thought it might be more helpful to say that it's not even a valid name (in case someone put `source-postgres` instead of `acmeCo/foo/source-postgres`). When we start pushing down more of this validation into SQL, it may make sense to treat this just as a "doesn't exist" error.